### PR TITLE
[SAIC-518] Migrates server group resources

### PR DESCRIPTION
**Pull request by timbyr**

### CODE REVIEW

**Comment by roman-verchikov** [comment]
I assume the patch is about server groups, not security groups?.. Please update commit message and pull request description accordingly

**Comment by roman-verchikov** [comment]
Please resolve pep8 issues:
```
./tests/cloudferrylib/os/compute/test_server_groups.py:73:80: E501 line too long (81 > 79 characters)
./tests/cloudferrylib/os/compute/test_server_groups.py:92:80: E501 line too long (81 > 79 characters)
./tests/cloudferrylib/os/compute/test_server_groups.py:99:80: E501 line too long (81 > 79 characters)
./tests/cloudferrylib/os/compute/test_server_groups.py:118:80: E501 line too long (81 > 79 characters)
./cloudferrylib/os/compute/server_groups.py:115:20: E126 continuation line over-indented for hanging indent
./cloudferrylib/os/compute/server_groups.py:119:16: E126 continuation line over-indented for hanging indent
```

File: cloudferrylib/os/actions/server_group_transporter.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+
+from cloudferrylib.base.action import transporter
+from cloudferrylib.os.compute.server_groups import Handler
+from cloudferrylib.utils import utils as utl
+
+LOG = utl.get_log(__name__)
+
+
+class ServerGroupTransporter(transporter.Transporter):
```
Comment (roman-verchikov) [inline cloudferrylib/os/actions/server_group_transporter.py]: Please add docstring describing the purpose of the task:
 - what it does (end-user overview)
 - how to use
 - required config options
 - dependent and required tasks

File: cloudferrylib/os/compute/server_groups.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+from novaclient import exceptions as nova_exc
+
+from cloudferrylib.base.resource import Resource
+from cloudferrylib.utils import utils as utl
+
+
+LOG = utl.get_log(__name__)
+
+
+class Handler(Resource):
```
Comment (roman-verchikov) [inline cloudferrylib/os/compute/server_groups.py]: Please add docstring
 - overview
 - how to use

Comment (roman-verchikov) [inline cloudferrylib/os/compute/server_groups.py]: The class name seems quite a bit too generic, given the number of SQL requests done in class methods, I guess it should be named differently and placed in a different module

**Comment by roman-verchikov** [comment]
@timbyr please update according to comments above

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.cloudferrylib.os.compute.test_server_groups
tests/cloudferrylib/os/compute/test_server_groups.py:45: [E1002(super-on-old-class), ServerGroupTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/compute/test_server_groups.py:106: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:107: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:110: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:111: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:125: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:126: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:129: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:136: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/1415/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.cloudferrylib.os.compute.test_server_groups
tests/cloudferrylib/os/compute/test_server_groups.py:45: [E1002(super-on-old-class), ServerGroupTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/compute/test_server_groups.py:109: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:110: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:113: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:114: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:129: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:130: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:133: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:140: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/1416/

**Comment by MirantisWorkloadMobility** [comment]
Functional test success for Grizzly to Icehouse migration

**Comment by MirantisWorkloadMobility** [comment]
Functional test success for Grizzly to Juno migration

**Comment by MirantisWorkloadMobility** [comment]
Functional test failure for Icehouse to Juno migration
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-functional-tests-icehouse-to-juno/422/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.cloudferrylib.os.compute.test_server_groups
tests/cloudferrylib/os/compute/test_server_groups.py:45: [E1002(super-on-old-class), ServerGroupTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/compute/test_server_groups.py:109: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:110: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:113: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:114: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:129: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:130: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:133: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:140: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class

**Comment by MirantisWorkloadMobility** [comment]
Ready

**Comment by miarmak** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.cloudferrylib.os.compute.test_server_groups
tests/cloudferrylib/os/compute/test_server_groups.py:45: [E1002(super-on-old-class), ServerGroupTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/compute/test_server_groups.py:109: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:110: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:113: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:114: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:129: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:130: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:133: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _delete_server_group of a client class
tests/cloudferrylib/os/compute/test_server_groups.py:140: [W0212(protected-access), ServerGroupTestCase.test_deploy_server_groups_already_exists_doesnt_match] Access to a protected member _deploy_server_group of a client class

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/os/compute/server_groups.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+SQL_SELECT_POLICY = ("SELECT policy FROM instance_group_policy WHERE "
+                     "group_id=%s AND deleted=0;")
+SQL_SELECT_GROUP_ID = "SELECT id FROM instance_groups WHERE uuid='{0}';"
+SQL_DELETE_MEMBER = "DELETE FROM instance_group_member WHERE group_id={0};"
+SQL_DELETE_POLICY = "DELETE FROM instance_group_member WHERE group_id={0};"
+SQL_DELETE_GROUP = "DELETE_FROM instance_groups WHERE uuid='{0}';"
+SQL_INSERT_GROUP = ("INSERT INTO instance_groups (uuid, name, project_id, "
+                    "user_id, deleted) VALUES('{0}', '{1}', '{2}', '{3}', 0);")
+SQL_INSERT_POLICY = ("INSERT INTO instance_group_policy (group_id, policy, "
+                     "deleted) VALUES({0}, '{1}', 0)")
```
Comment (roman-verchikov) [inline cloudferrylib/os/compute/server_groups.py]: Wow! More direct low-level DB modifications! Is there really no other way of dealing with server groups? These would require **very** thorough testing before merge.

Comment (timbyr) [inline cloudferrylib/os/compute/server_groups.py]: I'm still getting familiar with the existing cloud ferry code base. I think I can now do these with some of the existing utility functions and the APIs.

**Comment by miarmak** [comment]
Works on devlab and staging. Needs to be merged.

**Pull request merged**